qiime2/sdk/actiongraph.py

```diff
@@ -0,0 +1,217 @@
+# ----------------------------------------------------------------------------
+
+
+from itertools import combinations
+import networkx as nx
+import copy
+import qiime2
+
+
+def get_next_arguments(action, type=0):
+    """
+    Get a tuple of required/nonrequired inputs or outputs for each method
+
+    Parameters
+    ----------
+    action : Qiime2.action
+
+    type : {0,1,2}
+        Delineates if getting the action input, param, or output types
+
+    Returns
+    -------
+    Tuple - list of required and optional semantic types
+
+    """
+
+    req = []
+    non_req = []
+
+    if type == 0:
+        for k, v in action.signature.inputs.items():
+            if not v.has_default():
+                req.append(v.qiime_type)
+            else:
+                non_req.append(v.qiime_type)
+    elif type == 1:
+        for k, v in action.signature.parameters.items():
+            if not v.has_default():
+                req.append(v.qiime_type)
+            else:
+                non_req.append(v.qiime_type)
+    else:
+        for k, v in action.signature.outputs.items():
+            if not v.has_default():
+                req.append(v.qiime_type)
+            else:
+                non_req.append(v.qiime_type)
+
+    return req, non_req
+
+
+def get_combinations(no_req):
+    """
+    Get all permutations of list of semantic types
+
+    Parameters
+    ----------
+    non_req : list of semantic types
+
+    Returns
+    -------
+    List of lists - each list contains an instance of a permutation
+
+    """
+
+    if not no_req:
+        return no_req
+    no_req_comb = []
+    for i in range(len(no_req)+1):
+        combs = combinations(no_req, i)
+        for c in combs:
+            no_req_comb.append(list(c))
+    return no_req_comb
+
+
+def generate_nodes_by_action(action):
+    """
+    Given a method, generates all combinations of inputs and
+    outputs for that particular method and and stores the combinations
+    as dictionaries in a resulting list.
+
+    Parameters
+    ----------
+    method : Qiime2.action
+
+    Returns
+    -------
+    list of dictionaries - each dictionary is a combination inputs and
+        outputs for particular node
+
+    """
+
+    results = []
+
+    input, input_nr = get_next_arguments(action, 0)
+    param, param_nr = get_next_arguments(action, 1)
+    output, _ = get_next_arguments(action, 2)
+
+    non_req = []
+
+    # unravel potential unions
+    input = unravel_union(input)
+    param = unravel_union(param)
+
+    for x in input_nr + param_nr:
+        non_req += unravel_union(x)
+
+    for i in input:
+        for p in param:
+            if non_req:
+                non_req = get_combinations(non_req)
+                for nr in non_req:
+                    d = {}
+                    d['inputs'] = i
+                    d['params'] = p
+                    d['outputs'] = output
+                    d['non_req'] = nr
```

Comment:
**Change Requested**

This is 99% of what we're looking for, but the `non_req` section should be merged back into `inputs/params` depending on where the content of that came from.